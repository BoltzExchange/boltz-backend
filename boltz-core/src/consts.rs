pub const ECDSA_BYTES_TO_GRIND: usize = 2;

pub const PREIMAGE_DUMMY: [u8; 0] = [];

pub const STUB_SCHNORR_SIGNATURE_LENGTH: usize = 64;
