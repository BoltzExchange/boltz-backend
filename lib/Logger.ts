import winston from 'winston';
import colors from 'colors/safe';
import { getTsString } from './Utils';

class Logger {
  public static readonly disabledLogger = new Logger('', '', true);

  constructor(filename: string, level: string, private disabled = false) {
    if (disabled) {
      return;
    }

    winston.configure({
      level,
      transports: [
        new winston.transports.Console({
          format: this.getLogFormat(true),
        }),
        new winston.transports.File({
          filename,
          format: this.getLogFormat(false),
        }),
      ],
    });
  }

  private getLogFormat = (colorize: boolean) => {
    return winston.format.printf(info => `${getTsString()} ${this.getLevel(info.level, colorize)}: ${info.message}`);
  }

  private getLevel = (level: string, colorize: boolean) => {
    if (colorize) {
      switch (level) {
        case 'error': return colors.red(level);
        case 'warn': return colors.yellow(level);
        case 'info': return colors.green(level);
        case 'verbose': return colors.cyan(level);
        case 'debug': return colors.blue(level);
        case 'silly': return colors.magenta(level);
      }
    }
    return level;
  }

  public error = (message: string) => {
    this.log('error', message);
  }

  public warn = (message: string) => {
    this.log('warn', message);
  }

  public info = (message: string) => {
    this.log('info', message);
  }

  public verbose = (message: string) => {
    this.log('verbose', message);
  }

  public debug = (message: string) => {
    this.log('debug', message);
  }

  public silly = (message: string) => {
    this.log('silly', message);
  }

  private log = (level: string, message: string) => {
    if (!this.disabled) {
      winston.log(level, message);
    }
  }
}

export default Logger;
