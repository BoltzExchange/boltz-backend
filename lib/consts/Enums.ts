export enum ErrorCodePrefix {
  General = 0,
  Grpc = 1,
  Service = 2,
  Wallet = 3,
  Chain = 4,
  Lnd = 5,
  Swap = 6,
  Rates = 7,
  Backup = 8,
}

export enum ClientStatus {
  Disconnected,
  Connected,
  OutOfSync,
}

export enum OrderSide {
  BUY,
  SELL,
}

export enum SwapUpdateEvent {
  SwapCreated = 'swap.created',
  SwapExpired = 'swap.expired',

  ChannelCreated = 'channel.created',

  InvoiceSet = 'invoice.set',
  InvoicePaid = 'invoice.paid',
  InvoicePending = 'invoice.pending',
  InvoiceSettled = 'invoice.settled',
  InvoiceFailedToPay = 'invoice.failedToPay',

  TransactionFailed = 'transaction.failed',
  TransactionMempool = 'transaction.mempool',
  TransactionClaimed = 'transaction.claimed',
  TransactionRefunded = 'transaction.refunded',
  TransactionConfirmed = 'transaction.confirmed',

  // Events for the prepay miner fee Reverse Swap protocol
  MinerFeePaid = 'minerfee.paid',
}

export enum ChannelCreationType {
  Auto = 'auto',
  Create = 'create',
}

export enum ChannelCreationStatus {
  Attempted = 'attempted',
  Created = 'created',
  Settled = 'settled',
  Abandoned = 'abandoned',
}

export enum ServiceWarning {
  ReverseSwapsDisabled = 'reverse.swaps.disabled',
}

export enum ServiceInfo {
  PrepayMinerFee = 'prepay.minerfee',
}

export enum Network {
  Mainnet = 'mainnet',
  Testnet = 'testnet',
  Simnet = 'simnet',
  Regtest = 'regtest',
}

export enum SwapType {
  Submarine = 'submarine',
  ReverseSubmarine = 'reversesubmarine',
}
