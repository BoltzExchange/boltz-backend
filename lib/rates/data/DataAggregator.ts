import { getPairId } from '../../Utils';
import Exchange from './Exchange';
import Binance from './exchanges/Binance';
import Bitfinex from './exchanges/Bitfinex';
import CoinbasePro from './exchanges/CoinbasePro';
import Kraken from './exchanges/Kraken';

class DataAggregator {
  private readonly exchanges: Exchange[] = [
    new Kraken(),
    new Binance(),
    new Bitfinex(),
    new CoinbasePro(),
  ];

  public readonly pairs = new Set<[string, string]>();

  public latestRates = new Map<string, number>();

  public registerPair = (baseAsset: string, quoteAsset: string): void => {
    this.pairs.add([baseAsset, quoteAsset]);
  };

  public fetchPairs = async (): Promise<Map<string, number>> => {
    const rateMap = new Map<string, number>();

    const queryPromises: Promise<void>[] = [];

    const queryRate = async (base: string, quote: string) => {
      const pairId = getPairId({ base, quote });
      const rate = await this.getRate(base, quote);

      if (rate && !isNaN(rate)) {
        rateMap.set(pairId, rate);
      } else {
        // If the rate couldn't be fetched, the latest one should be used
        rateMap.set(pairId, this.latestRates.get(pairId) || NaN);
      }
    };

    this.pairs.forEach(([baseAsset, quoteAsset]) => {
      queryPromises.push(queryRate(baseAsset, quoteAsset));
    });

    await Promise.all(queryPromises);

    this.latestRates = rateMap;
    return rateMap;
  };

  private getRate = async (baseAsset: string, quoteAsset: string) => {
    const promises: Promise<number>[] = [];

    this.exchanges.forEach((exchange) =>
      promises.push(exchange.getPrice(baseAsset, quoteAsset)),
    );

    const results = await Promise.all(
      promises.map((promise) => promise.catch((error) => error)),
    );

    // Filter all results that are not numeric (failed requests)
    const validResults: number[] = results.filter(
      (result) => !isNaN(Number(result)),
    );
    validResults.sort((a, b) => a - b);

    const middle = (validResults.length - 1) / 2;

    if (validResults.length % 2 === 0) {
      return (
        (validResults[Math.ceil(middle)] + validResults[Math.floor(middle)]) / 2
      );
    } else {
      return validResults[middle];
    }
  };
}

export default DataAggregator;
