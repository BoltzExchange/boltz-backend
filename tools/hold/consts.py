from enum import Enum


class Network(str, Enum):
    Mainnet = "bitcoin"
    Testnet = "testnet"
    Signet = "signet"
    Regtest = "regtest"


PLUGIN_NAME = "hold"
VERSION = "0.0.2"

TIMEOUT_CANCEL = 60
TIMEOUT_CANCEL_REGTEST = 5

TIMEOUT_CHECK_INTERVAL = 10

GRPC_HOST_REGTEST = "0.0.0.0"  # noqa: S104
