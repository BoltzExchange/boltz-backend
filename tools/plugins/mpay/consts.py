PLUGIN_NAME = "mpay"
VERSION = "0.0.2"
