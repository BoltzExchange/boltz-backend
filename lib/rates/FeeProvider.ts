import Logger from '../Logger';
import { OrderSide } from '../consts/Enums';
import { PairConfig } from '../consts/Types';
import { mapToObject, getPairId, stringify, getChainCurrency, splitPairId } from '../Utils';

class FeeProvider {
  // A map between the symbols of the pairs and their percentage fees
  public percentageFees = new Map<string, number>();

  public static transactionSizes = {
    normalClaim: 170,

    reverseLockup: 153,
    reverseClaim: 138,
  };

  constructor(
    private logger: Logger,
    private getFeeEstimation: (symbol: string) => Promise<Map<string, number>>,
  ) {}

  public init = (pairs: PairConfig[]) => {
    pairs.forEach((pair) => {
      // Set the configured fee or fallback to 1% if it is not defined
      const percentage = pair.fee !== undefined ? pair.fee : 1;

      if (pair.fee === undefined) {
        this.logger.warn(`Setting default fee of ${percentage}% for pair ${pair.base}/${pair.quote} because none was specified`);
      }

      this.percentageFees.set(getPairId(pair), percentage / 100);
    });

    this.logger.debug(`Prepared data for fee estimations: ${stringify(mapToObject(this.percentageFees))}`);
  }

  public getFees = async (
    pair: string,
    rate: number,
    orderSide: OrderSide,
    amount: number,
    isReverse: boolean,
  ) => {
    let percentageFee = this.percentageFees.get(pair) || 0;

    if (percentageFee !== 0) {
      percentageFee = percentageFee * amount * rate;
    }

    const { base, quote } = splitPairId(pair);
    const chainCurrency = getChainCurrency(base, quote, orderSide, isReverse);

    return {
      percentageFee: Math.ceil(percentageFee),
      baseFee: await this.getBaseFee(chainCurrency, isReverse),
    };
  }

  public getBaseFee = async (chainCurrency: string, isReverse: boolean) => {
    const feeMap = await this.getFeeEstimation(chainCurrency);

    return this.calculateBaseFee(feeMap.get(chainCurrency)!, isReverse);
  }

  private calculateBaseFee = (satPerVbyte: number, isReverse: boolean) => {
    if (isReverse) {
      // The lockup transaction which spends a P2WPKH output (possibly more but we assume a best case scenario here),
      // locks up funds in a P2WSH swap and sends the change back to a P2WKH output has about 153 vbytes
      return satPerVbyte * FeeProvider.transactionSizes.reverseLockup;
    } else {
      // The claim transaction which spends a nested SegWit swap output and
      // sends it to a P2WPKH address has about 170 vbytes
      return satPerVbyte * FeeProvider.transactionSizes.normalClaim;
    }
  }
}

export default FeeProvider;
