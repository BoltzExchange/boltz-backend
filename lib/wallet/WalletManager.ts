import fs from 'fs';
import { providers } from 'ethers';
import { Network } from 'bitcoinjs-lib';
import { BIP32Interface, fromSeed } from 'bip32';
import { mnemonicToSeedSync, validateMnemonic } from 'bip39';
import Errors from './Errors';
import Wallet from './Wallet';
import Logger from '../Logger';
import { splitDerivationPath } from '../Utils';
import ChainClient from '../chain/ChainClient';
import LndClient from '../lightning/LndClient';
import { CurrencyType } from '../consts/Enums';
import KeyRepository from '../db/KeyRepository';
import EthereumManager from './ethereum/EthereumManager';
import ChainTipRepository from '../db/ChainTipRepository';
import { KeyProviderType } from '../db/models/KeyProvider';
import LndWalletProvider from './providers/LndWalletProvider';
import CoreWalletProvider from './providers/CoreWalletProvider';
import WalletProviderInterface from './providers/WalletProviderInterface';

type CurrencyLimits = {
  maxSwapAmount: number;
  minSwapAmount: number;

  minWalletBalance: number;

  minLocalBalance?: number;
  minRemoteBalance?: number;

  maxZeroConfAmount?: number;
};

type Currency = {
  symbol: string;
  type: CurrencyType,
  limits: CurrencyLimits;

  // Needed for UTXO based coins
  network?: Network;
  lndClient?: LndClient;
  chainClient?: ChainClient;

  // Needed for Ether and tokens on Ethereum
  provider?: providers.Provider;
};

/**
 * WalletManager creates wallets instances that generate keys derived from the seed and
 * interact with the wallet of LND to send and receive onchain coins
 */
class WalletManager {
  public wallets = new Map<string, Wallet>();

  public ethereumManager?: EthereumManager;

  private readonly mnemonic: string;
  private readonly masterNode: BIP32Interface;
  private readonly keyRepository: KeyRepository;

  private readonly derivationPath = 'm/0';

  constructor(private logger: Logger, mnemonicPath: string, private currencies: Currency[], ethereumManager?: EthereumManager) {
    this.mnemonic = this.loadMnemonic(mnemonicPath);
    this.masterNode = fromSeed(mnemonicToSeedSync(this.mnemonic));

    this.keyRepository = new KeyRepository();

    this.ethereumManager = ethereumManager;
  }

  /**
   * Initializes a new WalletManager with a mnemonic
   */
  public static fromMnemonic = (logger: Logger, mnemonic: string, mnemonicPath: string, currencies: Currency[], ethereumManager?: EthereumManager): WalletManager => {
    if (!validateMnemonic(mnemonic)) {
      throw(Errors.INVALID_MNEMONIC(mnemonic));
    }

    fs.writeFileSync(mnemonicPath, mnemonic);

    return new WalletManager(logger, mnemonicPath, currencies, ethereumManager);
  }

  public init = async (chainTipRepository: ChainTipRepository): Promise<void> => {
    const keyProviderMap = await this.getKeyProviderMap();

    for (const currency of this.currencies) {
      if (currency.type !== CurrencyType.BitcoinLike) {
        continue;
      }

      let walletProvider: WalletProviderInterface | undefined = undefined;

      // The LND client is also used as onchain wallet for UTXO based chains if available
      if (currency.lndClient !== undefined) {
        walletProvider = new LndWalletProvider(this.logger, currency.lndClient, currency.chainClient!);

      // Else the Bitcoin Core wallet is used
      } else {
        walletProvider = new CoreWalletProvider(this.logger, currency.chainClient!);

        // Sanity check that wallet support is compiled in
        try {
          await walletProvider.getBalance();
        } catch (error) {
          // No wallet support is compiled in
          if (error.message === 'Method not found') {
            throw Errors.NO_WALLET_SUPPORT(currency.symbol);
          } else {
            throw error;
          }
        }
      }

      let keyProviderInfo = keyProviderMap.get(currency.symbol);

      // Generate a new KeyProvider if that currency does not have one yet
      if (!keyProviderInfo) {
        keyProviderInfo = {
          highestUsedIndex: 0,
          symbol: currency.symbol,
          derivationPath: `${this.derivationPath}/${this.getHighestDepthIndex(keyProviderMap, 2) + 1}`,
        };

        keyProviderMap.set(currency.symbol, keyProviderInfo);

        await this.keyRepository.addKeyProvider({
          ...keyProviderInfo,
          symbol: currency.symbol,
        });
      }

      const wallet = new Wallet(
        this.logger,
        CurrencyType.BitcoinLike,
        walletProvider,
      );

      wallet.initKeyProvider(
        currency.network!,
        keyProviderInfo.derivationPath,
        keyProviderInfo.highestUsedIndex,
        this.masterNode,
        this.keyRepository,
      );

      this.wallets.set(currency.symbol, wallet);
    }

    if (this.ethereumManager) {
      const ethereumWallets = await this.ethereumManager.init(this.mnemonic, chainTipRepository);

      for (const [symbol, ethereumWallet] of ethereumWallets) {
        this.wallets.set(symbol, ethereumWallet);
      }
    }
  }

  private loadMnemonic = (filename: string) => {
    if (fs.existsSync(filename)) {
      return fs.readFileSync(filename, 'utf-8').trim();
    }

    throw(Errors.NOT_INITIALIZED());
  }

  private getKeyProviderMap = async () => {
    const map = new Map<string, KeyProviderType>();
    const keyProviders = await this.keyRepository.getKeyProviders();

    keyProviders.forEach((keyProvider) => {
      map.set(keyProvider.symbol, {
        symbol: keyProvider.symbol,
        derivationPath: keyProvider.derivationPath,
        highestUsedIndex: keyProvider.highestUsedIndex,
      });
    });

    return map;
  }

  private getHighestDepthIndex = (map: Map<string, KeyProviderType>, depth: number): number => {
    if (depth === 0) {
      throw(Errors.INVALID_DEPTH_INDEX(depth));
    }

    let highestIndex = -1;

    map.forEach((info) => {
      const split = splitDerivationPath(info.derivationPath);
      const index = split.sub[depth - 1];

      if (index > highestIndex) {
        highestIndex = index;
      }
    });

    return highestIndex;
  }
}

export default WalletManager;
export { Currency };
