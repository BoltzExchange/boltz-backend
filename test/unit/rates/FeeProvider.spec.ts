import Logger from '../../../lib/Logger';
import { BaseFeeType, OrderSide, SwapVersion } from '../../../lib/consts/Enums';
import FeeProvider from '../../../lib/rates/FeeProvider';
import DataAggregator from '../../../lib/rates/data/DataAggregator';
import WalletManager from '../../../lib/wallet/WalletManager';
import { Ethereum } from '../../../lib/wallet/ethereum/EvmNetworks';

const btcFee = 36;
const ltcFee = 3;
const ethFee = 11;

const getFeeEstimation = async () => {
  return new Map([
    ['BTC', btcFee],
    ['LTC', ltcFee],
    ['ETH', ethFee],
  ]);
};

jest.mock('../../../lib/rates/data/DataAggregator', () => {
  return jest.fn().mockImplementation(() => ({
    latestRates: new Map<string, number>([['ETH/USDT', 2]]),
  }));
});

const MockedDataAggregator = <jest.Mock<DataAggregator>>DataAggregator;

jest.mock('../../../lib/wallet/WalletManager', () => {
  return jest.fn().mockImplementation(() => ({
    ethereumManagers: [
      {
        networkDetails: Ethereum,
        hasSymbol: jest.fn().mockReturnValue(true),
      },
    ],
  }));
});

const MockedWalletManager = <jest.Mock<WalletManager>>WalletManager;

describe('FeeProvider', () => {
  const feeProvider = new FeeProvider(
    Logger.disabledLogger,
    MockedWalletManager(),
    MockedDataAggregator(),
    getFeeEstimation,
  );

  test('should init', () => {
    feeProvider.init([
      {
        base: 'LTC',
        quote: 'BTC',
        fee: 2,
        swapInFee: -1,
        minSwapAmount: 1,
        maxSwapAmount: 2,
      },
      {
        base: 'BTC',
        quote: 'BTC',
        fee: 0,
        swapInFee: -1,
        minSwapAmount: 1,
        maxSwapAmount: 2,
      },
      {
        base: 'LTC',
        quote: 'LTC',

        // The FeeProvider should set this to 1
        fee: undefined,
        swapInFee: undefined,

        minSwapAmount: 1,
        maxSwapAmount: 2,
      },
    ]);

    const feeMap = feeProvider['percentageFees'];
    expect(feeMap.size).toEqual(3);
  });

  test('should get percentage fees of a pair', () => {
    expect(feeProvider.getPercentageFees('LTC/BTC')).toEqual({
      percentage: 2,
      percentageSwapIn: -1,
    });
    expect(feeProvider.getPercentageFees('BTC/BTC')).toEqual({
      percentage: 0,
      percentageSwapIn: -1,
    });
    expect(feeProvider.getPercentageFees('LTC/LTC')).toEqual({
      percentage: 1,
      percentageSwapIn: 1,
    });
  });

  test('should get percentage fees of normal swaps', () => {
    expect(feeProvider.getPercentageFee('LTC/BTC', false)).toEqual(-0.01);
    expect(feeProvider.getPercentageFee('BTC/BTC', false)).toEqual(-0.01);

    // Should set undefined fees to 1%
    expect(feeProvider.getPercentageFee('LTC/LTC', false)).toEqual(0.01);
  });

  test('should get percentage fees of reverse swaps', () => {
    expect(feeProvider.getPercentageFee('LTC/BTC', true)).toEqual(0.02);
    expect(feeProvider.getPercentageFee('BTC/BTC', true)).toEqual(0);

    // Should set undefined fees to 1%
    expect(feeProvider.getPercentageFee('LTC/LTC', true)).toEqual(0.01);
  });

  test('should update miner fees', async () => {
    expect(feeProvider.minerFees.size).toEqual(0);

    await feeProvider.updateMinerFees('BTC');
    await feeProvider.updateMinerFees('LTC');
    await feeProvider.updateMinerFees('ETH');
    await feeProvider.updateMinerFees('USDT');

    expect(feeProvider.minerFees.size).toEqual(4);
  });

  test('should calculate miner fees', () => {
    expect(feeProvider.minerFees.get('BTC')).toEqual({
      [SwapVersion.Taproot]: {
        normal: 5436,
        reverse: {
          claim: 3996,
          lockup: 5544,
        },
      },
      [SwapVersion.Legacy]: {
        normal: 6120,
        reverse: {
          claim: 4968,
          lockup: 5508,
        },
      },
    });

    expect(feeProvider.minerFees.get('LTC')).toEqual({
      [SwapVersion.Taproot]: {
        normal: 453,
        reverse: {
          claim: 333,
          lockup: 462,
        },
      },
      [SwapVersion.Legacy]: {
        normal: 510,
        reverse: {
          claim: 414,
          lockup: 459,
        },
      },
    });

    expect(feeProvider.minerFees.get('ETH')).toEqual({
      [SwapVersion.Taproot]: {
        normal: 27416,
        reverse: {
          claim: 27416,
          lockup: 51106,
        },
      },
      [SwapVersion.Legacy]: {
        normal: 27416,
        reverse: {
          claim: 27416,
          lockup: 51106,
        },
      },
    });

    expect(feeProvider.minerFees.get('USDT')).toEqual({
      [SwapVersion.Taproot]: {
        normal: 53948,
        reverse: {
          claim: 53948,
          lockup: 191356,
        },
      },
      [SwapVersion.Legacy]: {
        normal: 53948,
        reverse: {
          claim: 53948,
          lockup: 191356,
        },
      },
    });
  });

  test('should get fees of a Swap', () => {
    const amount = 100000000;

    expect(
      feeProvider.getFees(
        'LTC/BTC',
        SwapVersion.Legacy,
        2,
        OrderSide.BUY,
        amount,
        BaseFeeType.NormalClaim,
      ),
    ).toEqual({
      baseFee: 6120,
      percentageFee: -2000000,
    });

    expect(
      feeProvider.getFees(
        'LTC/BTC',
        SwapVersion.Taproot,
        2,
        OrderSide.BUY,
        amount,
        BaseFeeType.NormalClaim,
      ),
    ).toEqual({
      baseFee: 5436,
      percentageFee: -2000000,
    });

    expect(
      feeProvider.getFees(
        'LTC/BTC',
        SwapVersion.Legacy,
        2,
        OrderSide.BUY,
        amount,
        BaseFeeType.ReverseLockup,
      ),
    ).toEqual({
      baseFee: 459,
      percentageFee: 4000000,
    });

    expect(
      feeProvider.getFees(
        'LTC/BTC',
        SwapVersion.Taproot,
        2,
        OrderSide.BUY,
        amount,
        BaseFeeType.ReverseLockup,
      ),
    ).toEqual({
      baseFee: 462,
      percentageFee: 4000000,
    });
  });

  test.each`
    symbol   | version                | type                         | expected
    ${'BTC'} | ${SwapVersion.Legacy}  | ${BaseFeeType.NormalClaim}   | ${6120}
    ${'BTC'} | ${SwapVersion.Taproot} | ${BaseFeeType.NormalClaim}   | ${5436}
    ${'BTC'} | ${SwapVersion.Legacy}  | ${BaseFeeType.ReverseClaim}  | ${4968}
    ${'BTC'} | ${SwapVersion.Taproot} | ${BaseFeeType.ReverseClaim}  | ${3996}
    ${'BTC'} | ${SwapVersion.Legacy}  | ${BaseFeeType.ReverseLockup} | ${5508}
    ${'BTC'} | ${SwapVersion.Taproot} | ${BaseFeeType.ReverseLockup} | ${5544}
    ${'LTC'} | ${SwapVersion.Legacy}  | ${BaseFeeType.NormalClaim}   | ${510}
  `(
    'should get base fees (symbol: $symbol, version: $version, type: $type)',
    ({ symbol, version, type, expected }) => {
      expect(feeProvider.getBaseFee(symbol, version, type)).toEqual(expected);
    },
  );

  test.each`
    gasPrice    | gasUsage   | expected
    ${0.065164} | ${100_000} | ${651}
    ${0.06}     | ${100_000} | ${600}
    ${0.042}    | ${100_000} | ${420}
    ${0.042}    | ${50_000}  | ${210}
    ${0.006}    | ${100_000} | ${60}
    ${0.0006}   | ${100_000} | ${6}
    ${0.0001}   | ${100_000} | ${1}
  `(
    'should calculate gas cost for gas price $gasPrice',
    ({ gasPrice, gasUsage, expected }) => {
      expect(FeeProvider['calculateEtherGasCost'](gasPrice, gasUsage)).toEqual(
        expected,
      );
    },
  );
});
