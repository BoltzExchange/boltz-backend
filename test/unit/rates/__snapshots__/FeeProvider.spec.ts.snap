// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeeProvider should honor discount CT flag of wallet 1`] = `
{
  "0": {
    "normal": 160,
    "reverse": {
      "claim": 166,
      "lockup": 301,
    },
  },
  "1": {
    "normal": 161,
    "reverse": {
      "claim": 158,
      "lockup": 301,
    },
  },
}
`;

exports[`FeeProvider should honor discount CT flag of wallet 2`] = `
{
  "0": {
    "normal": 26,
    "reverse": {
      "claim": 28,
      "lockup": 33,
    },
  },
  "1": {
    "normal": 22,
    "reverse": {
      "claim": 24,
      "lockup": 33,
    },
  },
}
`;
